### DISCUSSION TITLE
Stop addition steps after spindle stops

### DISCUSSION THREAD

**Discussion by rosspower** [kachurovskiy/nanoels#134] [Q&A]
Hello,
I purchased a h4 unit and love it, however, I have a pm1228 lathe and use it it drive the manual feeds.  The lathe has a 10:1 gearing in the feed so I set the pitch in gear mode to .08 in order to get a feed rate of .008".  As such, when the lathe spindle stops, the stepper continues for some time.  I would like to stop the addition steps when the spindle stops.  I tried using a buttonOn(false) but that requires the user to hit start every time the spindle starts.  How can I clear "waiting " steps when the spindle rpm =0?  Thanks in advance

**Comment by kachurovskiy** [comment]
Hi @rosspower, can you please share your modified code via gist.github.com and post a link here? Also please record a short video of the problem and attach to this question if you can - so that we understand the context better. Thanks!

**Reply by kachurovskiy** [reply]
Ross, taking a second look - it sounds like H4 doesn't have time to issue all the step pulses and this is why it's still running when the spindle stops. It does sound strange that it happens on such low feed rates though.

To solve this problem, try reducing microstepping on the motor as much as possible - ideally to 200 steps per revolution.

It would still help to see a short video of the problem and your modified h4.ino code.

**Comment by rosspower** [comment]
Hello, here is the code.  other than adding a melody when the unit turns on , the only thing I have done is to change the config my steps and such..... you might want to double check I got the math right.  the main lead screw is 8 TPI ( I have that in the config, double check it for me if you could.)  and I just added the X axis that is 10 TPI ( I have not set that up yet, but the controller is set to 1/8 step on a 200 step nema 17)  if you know that off hand let me know that as well.  I will post a video in another post

 here is the gist and a zip file as well
[https://gist.github.com/rosspower/83cf92bda66aea2674fb4bbbbff8c3fd](url)
[h4.zip](https://github.com/kachurovskiy/nanoels/files/13196976/h4.zip)

**Reply by kachurovskiy** [reply]
Controller can issue at most ~32000 steps per second.

You're running 0.08" = ~2mm pitch at 1000RPM = ~17RPS with 0.32mm leadscrew at 500 motor steps. This would require issuing `2 * 17 / 0.32 * 500 = 53125` steps per second - so it can't keep up.

To make it work, leadscrew needs to get "larger" (e.g. by reducing the reduction ratio) and motor steps need to be "reduced" (e.g. by switching from 800 steps per revolution to 200 steps on the driver).

Sadly it's not possible within the current code architecture to "stop the addition steps when the spindle stops" since encoder and motor positions are tightly related and change together. It's possible to go into the emergency stop when this situation occurs but I don't think this would be helpful.

**Comment by rosspower** [comment]
here is a quick vid

https://youtu.be/8DESwLX9nLc

**Comment by rosspower** [comment]
@kachurovskiy I changed out the reduction drive to a lower ration and got the step count down so that part is solved.  Thank you so much for your assistance.

I do have another question/request....  Is it possible that we can get the ELS to display/calculate on diameter as well as radius?  Most DRO have this function, it would be very helpful and greatly enhance the functionality of the unit.  Here are some examples:

I have a 12mm diameter stock, I need to turn it down to 10mm, I touch off the x axis on the diameter and set it to 0.  I then move the x axis in 2mm (12 -2 = 10) and set the inside x axis stop.  adjust the z axis start and stop as per normal (diameter/radius operation are only on the x axis)

in the example above, in diameter mode, the unit would display the 2mm move, but only move the x axis in by 1mm.  in radius mode, the user would have to adjust in by only 1mm, otherwise, after tuning, the stock would now be 8mm in diameter.

simple numbers like the above are easy to calc in your head to just move 1/2 and set what you want, however, others are not.

In order for us to add this functionality, we would need to adjust all displays and calculation regarding the x axis to be .5 of the movement displayed on all operations

Thoughts?

**Reply by kachurovskiy** [reply]
Ross, have you tried setting X0 from diameter? It's done by taking a light OD cut, measuring the resulting diameter, entering it in H4 numpad and pressing `A`. This sets X0 at `-D/2`. Please let me know if it helps.

**Comment by rosspower** [comment]
that is not quite what I mean or would like to see.  
so if I do what you say on a 12mm diameter stock.  touch off the tool and type 12 then the A, the display shows -6mm.  (like you said -d/2)
the issue is that if I advance the X in by 1mm on the display to -4mm and make a cut, what I have done is cut 1mm off radius thus taking the diameter down by 2mm and the part is now 10mm dia instead of 11.

in diameter mode on a standard DRO, when you move the tool in by .5mm the DRO indicates 1mm.  Likewise, if one were to move the x axis in by 1mm (as is on the H4, the DRO would show that as a 2mm movement.

Thoughts?

That is the function of a diameter/radius DRO  it swaps between diameter mode and radius mode.

**Reply by kachurovskiy** [reply]
> if I advance the X in by 1mm on the display to -4mm and make a cut, what I have done is cut 1mm off radius thus taking the diameter down by 2mm and the part is now 10mm dia instead of 11

Sorry if this caused problems. My goal with H4 is to keep things as simple as possible - in this case, logic is that X and Z position track the position of the cutter in a plane. I wouldn't want to break this abstraction yet. I think it's good to reduce the complexity and all kind of errors long term.

H4 doesn't have every feature that other products have (e.g. it's not possible to configure it via screen and it doesn't differentiate between feed and thread) but I hope that it provides plenty of functionality and is useful.

Supporting diameter mode on X is something that I will keep in mind but implementing it right is not trivial - axis position is taken/shown in many places - and it's easy to do more harm than good by confusing people. At the same time, I will also keep an eye on feedback from other users.
